src/system/mod.rs

```diff
@@ -35,19 +36,58 @@ pub mod term;
 
 pub mod wait;
 
-/// Mark all the file descriptors that are not the IO streams to be closed when `exec` is called.
-pub(crate) fn close_the_universe() -> io::Result<()> {
-    let min_fd = libc::STDERR_FILENO + 1;
-    let max_fd: c_int = -1;
-
-    cerr(unsafe {
-        libc::syscall(
-            libc::SYS_close_range,
-            min_fd,
-            max_fd,
-            libc::CLOSE_RANGE_CLOEXEC,
-        )
-    })?;
+/// A type able to close every file descriptor except for the ones pased via [`FileCloser::except`]
+/// and the IO streams.
+pub(crate) struct FileCloser {
+    fds: BTreeSet<c_uint>,
+}
+
+impl FileCloser {
+    pub(crate) const fn new() -> Self {
+        Self {
+            fds: BTreeSet::new(),
+        }
+    }
+
+    pub(crate) fn except<F: AsRawFd>(&mut self, fd: &F) {
+        self.fds.insert(fd.as_raw_fd() as c_uint);
+    }
+
+    /// Close every file descriptor that is not one of the IO streams or one of the file
+    /// descriptors passed via [`FileCloser::except`].
+    pub(crate) fn close_the_universe(self) -> io::Result<()> {
+        let mut fds = self.fds.into_iter();
+
+        let Some(mut curr_fd) = fds.next() else {
+            return close_range(STDERR_FILENO as c_uint + 1, c_uint::MAX);
+        };
+
+        if let Some(max_fd) = curr_fd.checked_sub(1) {
+            close_range(STDERR_FILENO as c_uint + 1, max_fd)?;
+        }
+
+        for next_fd in fds {
+            if let Some(min_fd) = curr_fd.checked_add(1) {
+                if let Some(max_fd) = next_fd.checked_sub(1) {
+                    close_range(min_fd, max_fd)?;
+                }
+            }
+
+            curr_fd = next_fd;
+        }
+
+        if let Some(min_fd) = curr_fd.checked_add(1) {
+            close_range(min_fd, c_uint::MAX)?;
+        }
+
+        Ok(())
+    }
+}
+
+fn close_range(min_fd: c_uint, max_fd: c_uint) -> io::Result<()> {
+    if min_fd <= max_fd {
```

Comment:
I think this should be an assertion because if the condition fails that likely signals a bug somewhere else

Reply:
I did this just because if you do:
```rust
closer.except(n);
closer.except(n + 1);
```
Then `closer.close_the_universe()` would call 
```rust
close_range((n) + 1, (n + 1) - 1);
```